src/compiler/checker.ts

```diff
@@ -27493,6 +27493,11 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         return type === unknownUnionType ? unknownType : type;
     }
 
+    // use to determine if a parameter may be undefined or null (or is unknown/unconstrained)
+    function getUnknownIfMaybeUnknown(type: Type) {
+        return (strictNullChecks && type.flags & TypeFlags.Instantiable) ? getBaseConstraintOfType(type) || unknownType : type;

```

Comment:
I'm assuming from #59059 that making this change to `getTypeFacts` breaks a lot of existing code, is that right?

Also, does making this change to `checkNonNullType` instead also break a lot of code?

Reply:
Yes, I tried that and putting this change direclty into `checkNonNullType` gets a very similar result to #59059 in terms of breaking object index checking

Reply:
A lot of the cases that break relate to this type of index access https://github.com/microsoft/TypeScript/pull/59059#issuecomment-2230097569

Reply:
I don't know if this it's noticeably different in semantics, but istead of doing this, you might want to try to modify `getBaseTypeOfLiteralTypeForComparison` and rename it `getBaseTypeForComparison`. Then pass the result of that into `checkNonNullType`.

Reply:
I would also consider renaming this to `getBaseConstraintOrUnknown`, use `??` instead of `||`, and just get rid of the `strictNullChecks` check.

Reply:
Also - any clue if this works on unions of unconstrained type parameters? Can you add a test?

```ts
function f<T, U>(x: T | U, y: T | U) {
  return x < y;
}
```